emukit/core/optimization/local_search_acquisition_optimizer.py

```diff
@@ -0,0 +1,110 @@
+                    neighbours.append(parameter.encodings[
+                        (parameter.encodings != features).any(axis=1)])
+                else:
+                    raise NotImplementedError("Parameter encoding {} not supported."
+                                              .format(type(parameter.encoding)))
+            else:
+                raise NotImplementedError("Parameter type {} not supported."
+                                          .format(type(parameter)))
+            current_feature += parameter.dimension
+        return neighbours
+
+    def _neighbours(self, all_features: np.ndarray) -> np.ndarray:
+        assert all_features.ndim == 1, "Expected 1d array, got {}d".format(all_features.ndim)
+
+        neighbours_per_param = self._neighbours_per_parameter(all_features)
+        num_neighbours = sum(param.shape[0] for param in neighbours_per_param)
+        neighbours = np.full((num_neighbours, all_features.shape[0]), all_features)
+        current_neighbour, current_feature = 0, 0
+        for this_neighbours in neighbours_per_param:
+            next_neighbour = current_neighbour + this_neighbours.shape[0]
+            next_feature = current_feature + this_neighbours.shape[1]
+            neighbours[current_neighbour:next_neighbour,
+                       current_feature:next_feature] = this_neighbours
+            current_neighbour, current_feature = next_neighbour, next_feature
+        return neighbours
+
+    def _one_local_search(self, acquisition: Acquisition, x: np.ndarray):
+        assert x.ndim == 1, "Expected 1d array, got {}d".format(x.ndim)
+
+        incumbent_value = acquisition.evaluate(x.reshape(1, -1)).item()
+        for step in range(self.num_steps):
+            _log.info("Step {}/{}".format(step + 1, self.num_steps))
+            neighbours = self._neighbours(x)
+            acquisition_values = acquisition.evaluate(neighbours)
+            max_index = np.argmax(acquisition_values)
+            max_neighbour = neighbours[max_index]
+            max_value = acquisition_values[max_index]
+            if max_value < incumbent_value:
+                _log.info("Converged")
+                break
+            else:
+                incumbent_value = max_value
+                x = max_neighbour
+        return x, incumbent_value
+
+    def optimize(self, acquisition: Acquisition, context: dict = None) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Optimizes the acquisition function, taking into account gradients if it supports them
+        :param acquisition: The acquisition function to be optimized
+        :param context: Optimization context.
+                        Determines whether any variable values should be fixed during the optimization
+        :return: Tuple of (location of maximum, acquisition value at maximizer)
+        """
+        if context is not None:
+            raise NotImplementedError("Handling context is currently not supported.")
```

Comment:
I don't think this warrants an error. A log warning max

Reply:
I think I disagree. I'd be kind of annoyed if I passed a context in, waited a while for the loop to converge then found my context was completely ignored.

Reply:
Perhaps, i can see a point in this too. Not sure what to do best, let me think more and get back to this point later

Reply:
Warning logs might get lost but your results change if context is (not) handled. Also it's more pythonic to be explicit here imo.  
If a user does not care about that much about the context, a fallback can be used.
```python
try:
    optimizer.optimize(acquisition, context)
except NotImplementedError:
     _log.warning("...")
    optimizer.optimize(acquisition, None)
```